fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java

```diff
@@ -0,0 +1,195 @@
+        if (delinquencyRange == null) {
+            throw DelinquencyRangeNotFoundException.notFound(delinquencyRangeId);
+        }
+        Map<String, Object> changes = new HashMap<>();
+        delinquencyRange = updateDelinquencyRange(delinquencyRange, data, changes);
+        return new CommandProcessingResultBuilder().withCommandId(command.commandId()).withEntityId(delinquencyRange.getId()).with(changes)
+                .build();
+    }
+
+    @Override
+    public CommandProcessingResult deleteDelinquencyRange(Long delinquencyRangeId, JsonCommand command) {
+        final DelinquencyRange delinquencyRange = repositoryRange.getReferenceById(delinquencyRangeId);
+        repositoryRange.delete(delinquencyRange);
+        return new CommandProcessingResultBuilder().withCommandId(command.commandId()).withEntityId(delinquencyRange.getId()).build();
+    }
+
+    @Override
+    public CommandProcessingResult createDelinquencyBucket(JsonCommand command) {
+        DelinquencyBucketData data = dataValidatorBucket.validateAndParseUpdate(command);
+        Map<String, Object> changes = new HashMap<>();
+        DelinquencyBucket delinquencyBucket = updateOrCreateDelinquencyBucket(data, changes);
+        return new CommandProcessingResultBuilder().withCommandId(command.commandId()).withEntityId(delinquencyBucket.getId()).with(changes)
+                .build();
+    }
+
+    @Override
+    public CommandProcessingResult updateDelinquencyBucket(Long delinquencyBucketId, JsonCommand command) {
+        DelinquencyBucketData data = dataValidatorBucket.validateAndParseUpdate(command);
+        DelinquencyBucket delinquencyBucket = this.repositoryBucket.getReferenceById(delinquencyBucketId);
+        if (delinquencyBucket == null) {
+            throw DelinquencyBucketNotFoundException.notFound(delinquencyBucketId);
+        }
+        Map<String, Object> changes = new HashMap<>();
+        delinquencyBucket = updateDelinquencyBucket(delinquencyBucket, data, changes);
+        return new CommandProcessingResultBuilder().withCommandId(command.commandId()).withEntityId(delinquencyBucket.getId()).with(changes)
+                .build();
+    }
+
+    @Override
+    public CommandProcessingResult deleteDelinquencyBucket(Long delinquencyBucketId, JsonCommand command) {
+        final DelinquencyBucket delinquencyBucket = repositoryBucket.getReferenceById(delinquencyBucketId);
+        repositoryBucket.delete(delinquencyBucket);
+        return new CommandProcessingResultBuilder().withCommandId(command.commandId()).withEntityId(delinquencyBucket.getId()).build();
+    }
+
+    @Override
+    @CronTarget(jobName = JobName.LOAN_DELINQUENCY_CLASSIFICATION)
+    public void setLoanDelinquencyClassification() throws JobExecutionException {
+
+    }
+
+    private DelinquencyRange updateOrCreateDelinquencyRange(DelinquencyRangeData data, Map<String, Object> changes) {
```

Comment:
I'm not really sure why these 2 cases have been implemented as one method.

Let's separate the responsibilities and have one method for creation and one for update.

Reply:
This is a similar of a recent implementation, BusinessDate to be more precise

Reply:
Guilty... :/